File: content/exercises/perfectNumbers/perfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

factors :: Int -> [Int]
factors n = [f | f<-[1..n], mod n f == 0]

classify :: Int -> Maybe Classification
classify n 
    | n < 1 = Nothing
    | 2*n == sumOfFactors = Just Perfect
    | 2*n > sumOfFactors = Just Deficient
    | otherwise = Just Abundant
    where sumOfFactors = sum $ factors n

File: content/exercises/perfectNumbers/perfectNumbers.py
from __future__ import annotations
from enum import Enum, auto


class Classify(Enum):
    DEFICIENT = auto()
    ABUNDENT = auto()
    PERFECT = auto()


def factors(n) -> list[int]:
    return set([f for f in range(1, n + 1) if n % f == 0])


def classify(n) -> Classify:
    sum_proper_factors = sum(factors(n)) - n
    if n == sum_proper_factors:
        return Classify.PERFECT
    if n > sum_proper_factors:
        return Classify.DEFICIENT
    return Classify.ABUNDENT
